lib/indexing/indexstore/indexstore.go

```diff
@@ -0,0 +1,359 @@
+					batch.Entries = make([]gocql.BatchEntry, 0, i.settings.InsertBatchSize)
+				}
+
+				batch.Entries = append(batch.Entries, gocql.BatchEntry{
+					Stmt:       Qry,
+					Args:       []any{pieceCidBytes, rec.Cid.Hash(), rec.Offset, rec.Size},
+					Idempotent: true,
+				})
+
+				if allIdx == len(recsb)-1 || len(batch.Entries) == i.settings.InsertBatchSize {
+					err := func() error {
+						defer func(start time.Time) {
+							log.Debugw("addIndex Batch Insert", "took", time.Since(start), "entries", len(batch.Entries))
+						}(time.Now())
+
+						err := i.session.ExecuteBatch(batch)
+						if err != nil {
+							return fmt.Errorf("executing batch insert for piece %s: %w", pieceCid, err)
+						}
+						return nil
+					}()
+					if err != nil {
+						return err
+					}
+					batch = nil
+				}
+			}
+			return nil
+		})
+	}
+
+	err := eg.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveIndexes removes all multihash -> piece cid mappings, and all
+// offset / size information for the piece.
+func (i *IndexStore) RemoveIndexes(ctx context.Context, pieceCid cid.Cid) error {
+	Qry := `DELETE FROM PayloadToPiece WHERE PayloadMultihash = ? AND PieceCid = ?`
+	pieceCidBytes := pieceCid.Bytes()
+
+	// Get multihashes for piece
+	recs, err := i.GetIndex(ctx, pieceCid)
+	if err != nil {
+		return fmt.Errorf("removing indexes for piece %s: getting recs: %w", pieceCid, err)
+	}
+
+	// Batch delete with concurrency
+	var eg errgroup.Group
+	for j := 0; j < i.settings.InsertConcurrency; j++ {
+		eg.Go(func() error {
+			var batch *gocql.Batch
+			var num int
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case rec, ok := <-recs:
+				if !ok {
+					// Finished adding all the queued items, exit the thread
+					err := i.session.ExecuteBatch(batch)
+					if err != nil {
+						return fmt.Errorf("executing batch delete for piece %s: %w", pieceCid, err)
+					}
+					return nil
+				}
+
+				if batch == nil {
+					batch = i.session.NewBatch(gocql.UnloggedBatch).WithContext(ctx)
+					batch.Entries = make([]gocql.BatchEntry, 0, i.settings.InsertBatchSize)
+				}
+
+				batch.Entries = append(batch.Entries, gocql.BatchEntry{
+					Stmt:       Qry,
+					Args:       []any{pieceCidBytes, rec.Cid.Hash(), rec.Offset, rec.Size},
+					Idempotent: true,
+				})
+
+				num++
+
+				if num == i.settings.InsertBatchSize {
+					err := i.session.ExecuteBatch(batch)
+					if err != nil {
+						return fmt.Errorf("executing batch delete for piece %s: %w", pieceCid, err)
+					}
+				}
+
+			}
+			return nil
+		})
+	}
+	err = eg.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type IndexRecord struct {
+	Record
+	Error error `json:"error,omitempty"`
+}
+
+// GetIndex retrieves the multihashes and offset/size information for a given piece CID.
+// It returns a channel of `IndexRecord` structs, which include the CID and the offset/size information.
```

Comment:
may be a better fit to return an iterator. above too.

Reply:
I am probably going to get rid of GetIndex func as it serves no purpose anymore. The records don't have full CID so they can only be used to match a provided CID.